packages/ai-jsx/src/next/index.ts

```diff
@@ -0,0 +1,114 @@
+import 'server-only';
+import * as ReactModule from 'react';
+import * as LLMx from '../react/core.js';
+export * from '../react/core.js';
+import { markAsJsxBoundary } from '../react/jsx-boundary.js';
+
+function unwrapReact(partiallyRendered: LLMx.PartiallyRendered): ReactModule.ReactNode {
+  if (LLMx.isElement(partiallyRendered)) {
+    // This should be an AI.React element.
+    if (partiallyRendered.tag !== LLMx.React) {
+      throw new Error('unwrapReact only expects to see AI.React elements or strings.');
+    }
+
+    return partiallyRendered.props.children;
+  }
+
+  return partiallyRendered;
+}
+
+function computeSuffix(prefix: ReactModule.ReactNode[], frame: ReactModule.ReactNode[]) {
+  const computedSuffix = [] as ReactModule.ReactNode[];
+  for (let i = 0; i < frame.length; ++i) {
+    if (i >= prefix.length) {
+      computedSuffix.push(frame[i]);
+    } else if (computedSuffix.length > 0) {
+      return null;
+    } else if (prefix[i] === frame[i]) {
+      continue;
+    } else {
+      const fromPrefix = prefix[i];
+      const fromFrame = frame[i];
+
+      if (typeof fromPrefix === 'string' && typeof fromFrame === 'string' && fromFrame.startsWith(fromPrefix)) {
+        computedSuffix.push(fromFrame.slice(fromPrefix.length));
+      } else {
+        return null;
+      }
+    }
+  }
+
+  return computedSuffix;
+}
+
+/**
+ * A JSX component that allows AI.jsx elements to be used in an RSC component tree.
+ */
+export function jsx({ children }: { children: LLMx.Node }, context?: any | LLMx.ComponentContext) {
+  if (typeof context?.render === 'function') {
+    // We're in AI.JSX already.
+    return children;
+  }
+
+  const renderResult = LLMx.createRenderContext().render(children, {
+    stop: (e) => e.tag === LLMx.React,
+    map: (frame) => frame.map(unwrapReact),
+  });
+  const asyncIterator = renderResult[Symbol.asyncIterator]();
+
+  /**
+   * (Ab)uses <Suspense> to stream an AI.JSX response, taking care to only stream deltas if
+   * each frame only appends to the previous one.
+   */
+  async function Stream({
```

Comment:
So we don't need to synchronously return all `Suspense` elements up-front? That was the presumption behind my shitty hack. :smile:

Reply:
Yeah, you can nest `<Suspense>` elements and Next takes care of making sure that an in-progress async RSC suspends the nearest ancestor. :)

Reply:
That's cool – RSC is even better than I gave it credit for!

Where did you learn about that behavior?